File: Scala3/src/test/scala/parser/ParserTest.scala
package parser

import lexer.{Lexer, Scanner, Token}
import org.junit.Assert.{assertEquals, assertTrue}
import org.junit.Test

import scala.util.Success

class ParserTest:
  @Test def parse(): Unit =
    val scanner = Scanner("""[null,true,false,[0,"あabc",-23.2],{"name": {"no":10e2}}]""")

    val array = List(
      JsonValue.Null,
      JsonValue.True,
      JsonValue.False,
      JsonValue.Array(
        List(
          JsonValue.Number(0.0),
          JsonValue.String("あabc"),
          JsonValue.Number(-23.2),
        )
      ),
      JsonValue.Object(
        Map(
          "name" -> JsonValue.Object(
            Map("no" -> JsonValue.Number(1000.0))
          )
        )
      )
    )

    assertEquals(Success(JsonValue.Array(array)), Parser.parse(scanner))

  @Test def parse_empty(): Unit =
    val scanner = Scanner("")
    assertEquals(Success(JsonValue.Null), Parser.parse(scanner))

  @Test def parse_lf(): Unit =
    val scanner = Scanner("\n")
    assertEquals(Success(JsonValue.Null), Parser.parse(scanner))

  @Test def parse_number(): Unit =
    val scanner = Scanner("-123.45")
    assertEquals(Success(JsonValue.Number(-123.45)), Parser.parse(scanner))

  @Test def parse_string(): Unit =
    val scanner = Scanner(""""abc"""")
    assertEquals(Success(JsonValue.String("abc")), Parser.parse(scanner))

  @Test def parse_empty_array(): Unit =
    val scanner = Scanner("[]")
    assertEquals(Success(JsonValue.Array(List())), Parser.parse(scanner))

  @Test def parse_array(): Unit =
    val scanner = Scanner("""[1, true, "a"]""")
    val array = List(JsonValue.Number(1.0), JsonValue.True, JsonValue.String("a"))
    assertEquals(Success(JsonValue.Array(array)), Parser.parse(scanner))

  @Test def parse_empty_object(): Unit =
    val scanner = Scanner("{}")
    assertEquals(Success(JsonValue.Object(Map())), Parser.parse(scanner))

  @Test def parse_object(): Unit =
    val scanner = Scanner("""{"key1": true, "key2": 123}""")
    val map = Map("key1" -> JsonValue.True, "key2" -> JsonValue.Number(123.0))
    assertEquals(Success(JsonValue.Object(map)), Parser.parse(scanner))

File: Java/app/src/test/java/com/github/shin1x1/jsonparser/parser/ParserTest.java
package com.github.shin1x1.jsonparser.parser;

import com.github.shin1x1.jsonparser.lexer.Lexer;
import com.github.shin1x1.jsonparser.lexer.Scanner;
import com.github.shin1x1.jsonparser.lexer.exception.UnexpectedEotException;
import com.github.shin1x1.jsonparser.parser.exception.UnexpectedTokenException;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.io.IOException;
import java.math.BigDecimal;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ParserTest {
    @Test
    void parse_number() throws IOException {
        var scanner = new Scanner("123.45");
        var sut = new Parser(new Lexer(scanner));

        assertEquals(new JsonValue.Number("123.45"), sut.parse());
    }

    @Test
    void parse_string() throws IOException {
        var scanner = new Scanner("\"abc\"");
        var sut = new Parser(new Lexer(scanner));

        assertEquals(new JsonValue.String("abc"), sut.parse());
    }

    @Test
    void parse_literal() throws IOException {
        var scanner = new Scanner("true");
        var sut = new Parser(new Lexer(scanner));

        assertEquals(new JsonValue.True(), sut.parse());
    }

    @Test
    void parse_empty() throws IOException {
        var scanner = new Scanner("");
        var sut = new Parser(new Lexer(scanner));

        assertEquals(new JsonValue.Null(), sut.parse());
    }

    @Test
    void parse_array() throws IOException {
        var scanner = new Scanner("[1,true,\"a\"]");
        var sut = new Parser(new Lexer(scanner));

        List<JsonValue> expectedList = Arrays.asList(
                new JsonValue.Number("1"),
                new JsonValue.True(),
                new JsonValue.String("a")
        );
        assertEquals(new JsonValue.Array(expectedList), sut.parse());
    }


    @Test
    void parse_array_recursive() throws IOException {
        var scanner = new Scanner("[[0,-1,3],\"a\"]");
        var sut = new Parser(new Lexer(scanner));

        List<JsonValue> expectedList = Arrays.asList(
                new JsonValue.Array(
                        Arrays.asList(
                                new JsonValue.Number("0"),
                                new JsonValue.Number("-1"),
                                new JsonValue.Number("3")
                        )
                ),
                new JsonValue.String("a")
        );
        assertEquals(new JsonValue.Array(expectedList), sut.parse());
    }

    @Test
    void parse_object() throws IOException {
        var scanner = new Scanner("{\"key1\": 1, \"key2\": {\"a\": [1,2,3]}}");
        var sut = new Parser(new Lexer(scanner));

        var expectedMap = new HashMap<String, JsonValue>() {
            {
                put("key1", new JsonValue.Number(new BigDecimal("1")));
                put("key2", new JsonValue.Object(new HashMap<>() {
                    {
                        put("a", new JsonValue.Array(Arrays.asList(
                                new JsonValue.Number("1"),
                                new JsonValue.Number("2"),
                                new JsonValue.Number("3")
                        )));
                    }
                }));
            }
        };
        assertEquals(new JsonValue.Object(expectedMap), sut.parse());
    }

    @ParameterizedTest
    @ValueSource(strings = {"{:1}", "{{", "[,", ":"})
    void parse_UnexpectedToken(String json) throws IOException {
        var scanner = new Scanner(json);
        var sut = new Parser(new Lexer(scanner));

        assertThrows(UnexpectedTokenException.class, sut::parse);
    }

    @ParameterizedTest
    @ValueSource(strings = {"[", "[1", "{", "{\"k\":", "\"a"})
    void parse_UnexpectedEot(String json) throws IOException {
        var scanner = new Scanner(json);
        var sut = new Parser(new Lexer(scanner));

        assertThrows(UnexpectedEotException.class, sut::parse);
    }
}